### PR TITLE
[AT-40][Add] Preliminary Logging CorrelationID

### PR DESCRIPTION
**Pull request by martinstibbe**

### CODE REVIEW

File: akamai/util.go [hunk 1]
```diff
@@ -97,3 +103,45 @@ func makeCache() *cache.Cache {
...
+func PrintLogFooter() {
+	edge.LogCorrelationID = KeepLogCorrelationID
+	if edge.LogCorrelationID != nil {
+		strLogCorrelationIDValue := *edge.LogCorrelationID
+		log.Printf("[DEBUG] END CORRELATION ID %s\n", strLogCorrelationIDValue)
+		edge.LogCorrelationID = nil
+	}
+}
+
+func CreateNonce() string {
```
Comment (zstlaw) [inline akamai/util.go:140]: I feel the nonce generation should be in the open-golang library as a common function so that it can be used in all levels of code but there may be reasons that is not advisable.

File: akamai/data_akamai_cp_code.go [hunk 2]
```diff
@@ -30,6 +31,8 @@ func dataSourceCPCode() *schema.Resource {
 }
 
 func dataSourceCPCodeRead(d *schema.ResourceData, meta interface{}) error {
+	setCorrelationID("dataSourceCPCodeRead-" + CreateNonce())
+	PrintLogHeader()
```
Comment (zstlaw) [inline akamai/data_akamai_cp_code.go]: This logging approach - header and footer log messages wrapping the actual message - unnecessarily increases the log size while also having a problem with concurrency and parallelism making the logs hard to parse and identify lines with the guid. We can not filter on the key since the rows we want are between the start and end tags instead of including them.  
 
A better solution is pushing a GUID tag into individual log statements and/or adopting a log statement pattern convention. I do not want a customer to cut parallelism to one just to achieve readability of the correlations in the log file. 
 
The a couple different ways you can do this :

1. Standardize the format of log lines.  example:
log.Printf(“[%s] [%s] [%s] %s”, logLevel, api, guid, msg)
e.g. log.Printf(“[DEBUG] [GTMv1] [%s] %s”, "DEBUG", "GTMv1", guid, "Opening call to server") with a guid that is a hash or a string like 19872984-be13-321e-a921731dbe9
This would result in a log like like the following 
[DEBUG] [GTMv1] [19872984-be13-321e-a921731dbe9] Opening call to server
2. Create a class to wrap our logging functions to helps provide the formatting above but have method signatures closer to printf in syntax so it is easier to plug in.  It would need to accept the range of formats supported by printf at least and then would build a new logging string.  Problem is this approach makes it harder to append guid later in the string and is more likely to look like "[guid] [level] [api] msg" since it is far safer to add a new tag at front than it is to edit strings of unknown format.
3. Create a class to standardize logging but instead of making it mirror printf you just rely on replacing existing printf statements to migrate to the new approach.  This is probably the safest approach as it allows us to get the formatting of choice 1 above and groups all the logging code together to make reuse possible.  But eliminates the overhead of making it a plug in replacement for printf relying instead on upfront work to migrate to the new approach.

File: akamai/util.go [hunk 3]
```diff
@@ -97,3 +103,45 @@ func makeCache() *cache.Cache {
 	profilecache := cache.New(5*time.Minute, 10*time.Minute)
 	return profilecache
 }
+
+func setCorrelationID(correlationid string) {
+	//edge.SetupLogging()
+	//log.Printf("[DEBUG] setCorrelationID %s\n", correlationid)
+	edge.SetLogCorrelationId(correlationid)
```
Comment (zstlaw) [inline akamai/util.go]: Won't concurrent calls to this by parallel resources overwrite each other's values in edge?

**Review by zstlaw — CHANGES_REQUESTED** [review]
Header and footer approach used here leave correlation to be done by hand at later point when debugging.  It would be far better to include guid in log statements and perhaps even require guid when calling openEdge-golang functions so they can log with the same guid generated in the resource/data endpoint because it is passed into the call.  (See extensive comment on the first occurrence of the logging in the dataSourceCPCodeRead method.

File: akamai/util.go (hunk 3 above)
Comment (zstlaw) [reply akamai/util.go]: I tested and because a shared variable is used the end tags are wrong: (trimmed example to shore the core problem)

START CORRELATION ID 305
Reading CP Code
START CORRELATION ID dd0
Reading CP Code
START CORRELATION ID 103
Reading CP Code
data.akamai_cp_code.example3:cpc_309358
END CORRELATION ID 103
data.akamai_cp_code.example2:cpc_23506
END CORRELATION ID 103

(third end statement was missing because shared variable had already been cleared)

File: akamai/data_akamai_cp_code.go (hunk 2 above)
Comment (ssrirama-aka) [reply akamai/data_akamai_cp_code.go]: I disagree with the comment that the header/footer approach unnecessarily increases log size. Log size was bound to increase incrementally with the introduction of correlations. But I agree that header/footer does not fit the parallelism inherent to the provider and TF. 

Once I generated logs with this branch, I found that reading off the header/footer information versus the log statements themselves to be counter-intuitive. As a developer, I am used to seeing the key fields of a log statement in the statement itself, going from left to right. So agree that introducing correlation ID into the log statement is a better convention.

I also agree that we need to define a log statement convention as you describe with #1. I would still place [level][api][guid]<message> in that order. The #2 option is not incompatible with #1, in the sense that you describe a Java LogAppender style implementation, to which the log format can be fed in as  regex.

Question is whether Hashicorp / Terraform providers follow a particular convention for log statements or its entirely up to individual providers to choose.

Comment (ssrirama-aka) [reply akamai/data_akamai_cp_code.go]: The one drawback I see of the log statement embedded GUID approach is that there will no longer be an explicit demarcation of start/finish or header/footer, but a grouping of a set of log statements. Editors' search functions would have to be used to find the first and last occurrence. I expect users will get used to it

Comment (martinstibbe) [reply akamai/data_akamai_cp_code.go]: I would prefer prefix as it makes it easy to cat logifile.txt|grep GUID|remove GUID prefix and you have log file raw and sequential also if we used common logging mechanism that is used in TF provider Debug content would be in same TF log file versus in 2 places (I used to do this in feature branch to test AI calls for bugs)

File: akamai/util.go (hunk 3 above)
Comment (martinstibbe) [reply akamai/util.go]: Each resource apply another thread so if we can get variable to be per instance of call should hold up better (was just trying to get that start end mark to work easier to mark lines I would think)

Comment (zstlaw) [reply akamai/util.go]: Making it per instance would make it work better but the interleaved calls mean that inline guid is still more efficient way to track a failed call for example.

**Review by zstlaw — APPROVED** [review]
Like this approach much better for clarity and readability of log files.
